### PR TITLE
fix(firebase_analytics): support complex data structures like list and map on Android

### LINKED ISSUE #4123: 🐛 [firebase_analytics] Item relevant events do not sent the right properties
## Bug report

For functions logging specific events involving items (like for example `logViewItem`, `logAddToCart` or `logRemoveFromCart`) the parameters are not sent in [the format Google Analytics expects](https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics.Event#public-static-final-string-view_item).

This means that among other things explorations involving `item_name` in Google Analytics fail.

### Expected behaviour

The above mentioned events are logged [as documented](https://firebase.google.com/docs/reference/android/com/google/firebase/analytics/FirebaseAnalytics.Event#public-static-final-string-view_item)

## Additional context

This should just show the issue more directly, currently I do not have the time to do a thorough PR myself:

Currently `logViewItem` (and similarly the other fuctions) is implemented as follows:

```dart
Future<void> logViewItem({
    @required String itemId,
    @required String itemName,
    @required String itemCategory,
    String itemLocationId,
    double price,
    int quantity,
    String currency,
    double value,
    String flightNumber,
    int numberOfPassengers,
    int numberOfNights,
    int numberOfRooms,
    String origin,
    String destination,
    String startDate,
    String endDate,
    String searchTerm,
    String travelClass,
  }) {
    _requireValueAndCurrencyTogether(value, currency);

    return logEvent(
      name: 'view_item',
      parameters: filterOutNulls(<String, dynamic>{
        _ITEM_ID: itemId,
        _ITEM_NAME: itemName,
        _ITEM_CATEGORY: itemCategory,
        _ITEM_LOCATION_ID: itemLocationId,
        _PRICE: price,
        _QUANTITY: quantity,
        _CURRENCY: currency,
        _VALUE: value,
        _FLIGHT_NUMBER: flightNumber,
        _NUMBER_OF_PASSENGERS: numberOfPassengers,
        _NUMBER_OF_NIGHTS: numberOfNights,
        _NUMBER_OF_ROOMS: numberOfRooms,
        _ORIGIN: origin,
        _DESTINATION: destination,
        _START_DATE: startDate,
        _END_DATE: endDate,
        _SEARCH_TERM: searchTerm,
        _TRAVEL_CLASS: travelClass,
      }),
    );
  }
```

but if `itemId` and `itemName` should be logged properly the call would actually be

```dart
logEvent(name: 'view_item', parameters: {
      'items': [
        {
          'item_id': itemId,
          'item_name': itemName,
        },
      ],
    });
```

### PR DESCRIPTION
**Pull request by lColinDl**
## Description

The native Android implementation of the firebase_analytics plugin does not support passing complex data structures (e.g. lists) as parameters into the `logEvent`method. This is for example necessary for ecommerce events like "view_item_list".
The native plugin implementation for iOS supports this already.

This PR is based on this already open, but outdated PR (https://github.com/FirebaseExtended/flutterfire/pull/1458). I don't care which PR gets merged, but as this is blocking the usage of ecommerce events, this is really important for me.

## Related Issues

This PR fixes:
- Fixes https://github.com/FirebaseExtended/flutterfire/issues/1036
- Fixes https://github.com/FirebaseExtended/flutterfire/issues/2835
- Fixes #4123 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by lColinDl** [comment]
@googlebot I fixed it.

**Comment by muccy** [comment]
Sorry for the bump, but this seems critical to me and I was about to make a similar PR

**Comment by CameliaTripon** [comment]
When can we expect this to be merged? We really need this fix for our application

**Comment by russellwheatley** [comment]
Hey @lColinDl, it would be great if you could have add a test to confirm an array and map type works [here](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart#L34) for us to consider this. Something like:
```dart
await FirebaseAnalytics().logEvent(
      name: 'test_complex_data',
      parameters: <String, dynamic>{
        'items': []// complex types here
      },
    )

```

 I wasn't able to log complex data types using your changes in my own preliminary investigation. It was an invalid type error from the native side.

**Comment by rrousselGit** [comment]
Agreed with Russell, we'll want tests.

I also believe we'll want iOS and web support too.

**Comment by lColinDl** [comment]
iOS worked out of the box and if I look at the code for web it should also work. It is just Android that was not working. I wrote the test for the change now in the same way as the existing tests. Hope its fine now. I'm not sure why it's not working for you @russellwheatley and what exactly you tested, but the implementation works for me for all enhanced ecommerce events that I had to implement by now. If there is a specific case that is not working, please specify it for me, so I can test and adjust the code.

File: packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart [hunk 1]
```diff
@@ -30,5 +30,23 @@ void main() {
     expect(
         FirebaseAnalytics().setCurrentScreen(screenName: 'testing'), completes);
     expect(FirebaseAnalytics().logEvent(name: 'testing'), completes);
```
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart:32]: Could you add a test for maps inside maps?

Comment (russellwheatley) [reply packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart:32]: great stuff so far, @lColinDl 🎉 . I agree with @rrousselGit, we should probably test a nested map as well 👍

Comment (rrousselGit) [reply packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart:32]: Thanks!

Could you also add a test for the `throw IllegalArgumentException`?

**Comment by Salakar** [comment]
Just waiting on the below then I think we're good to merge: 

image

File: packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart (hunk 1 above)
Comment (lColinDl) [reply packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart:35]: Done

**Review by rrousselGit — DISMISSED** [review]

**Comment by Salakar** [comment]
Looks like there's a failing Web test, could you take a look please? Thanks?

File: packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart [hunk 2]
```diff
@@ -32,5 +33,39 @@ void main() {
...
+                  'detail_1': 1,
+                  'detail_2': '2',
+                },
+              },
+            ],
+          },
+        ),
+        completes);
+
+    if (defaultTargetPlatform == TargetPlatform.android && !kIsWeb) {
```
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/example/test_driver/firebase_analytics_e2e.dart:59]:
```suggestion
    if (!kIsWeb && defaultTargetPlatform == TargetPlatform.android) {
```

**Review by rrousselGit — APPROVED** [review]

**Review by russellwheatley — APPROVED** [review]
LGTM

**Pull request merged**